react/features/settings/components/web/audio/AudioSettingsContent.js

```diff
@@ -0,0 +1,254 @@
+     * Click handler for the microphone entries.
+     *
+     * @param {string} deviceId - The deviceId for the clicked microphone.
+     * @returns {void}
+     */
+    _onMicrophoneEntryClick(deviceId) {
+        this.props.setAudioInputDevice(deviceId);
+    }
+
+    _onSpeakerEntryClick: (string) => void;
+
+    /**
+     * Click handler for the speaker entries.
+     *
+     * @param {string} deviceId - The deviceId for the clicked speaker.
+     * @returns {void}
+     */
+    _onSpeakerEntryClick(deviceId) {
+        this.props.setAudioOutputDevice(deviceId);
+    }
+
+    /**
+     * Renders a single microphone entry.
+     *
+     * @param {Object} data - An object with the deviceId and label of the microphone.
+     * @param {number} index - The index of the element, used for creating a key.
+     * @returns {React$Node}
+     */
+    _renderMicrophoneEntry(data, index) {
+        const { deviceId, label } = data;
+        const key = `me-${index}`;
+        const isSelected = deviceId === this.props.currentMicDeviceId;
+        let jitsiTrack = null;
+        let hasError = false;
+
+        if (isSelected) {
+            ({ jitsiTrack, hasError } = this.state.currentMicData);
+        }
+
+        return (
+            <MicrophoneEntry
+                deviceId = { deviceId }
+                hasError = { hasError }
+                isSelected = { isSelected }
+                jitsiTrack = { jitsiTrack }
+                key = { key }
+                onClick = { this._onMicrophoneEntryClick }>
+                {label}
+            </MicrophoneEntry>
+        );
+    }
+
+    /**
+     * Renders a single speaker entry.
+     *
+     * @param {Object} data - An object with the deviceId and label of the speaker.
+     * @param {number} index - The index of the element, used for creating a key.
+     * @returns {React$Node}
+     */
+    _renderSpeakerEntry(data, index) {
+        const { deviceId, label } = data;
+        const key = `se-${index}`;
+
+        return (
+            <SpeakerEntry
+                deviceId = { deviceId }
+                isSelected = { deviceId === this.props.currentOutputDeviceId }
+                key = { key }
+                onClick = { this._onSpeakerEntryClick }>
+                {label}
+            </SpeakerEntry>
+        );
+    }
+
+    /**
+     * Disposes the currently used audio track.
+     *
+     * @returns {Promise<void>}
+     */
+    _disposeCurrentTrack() {
+        const {
+            currentMicData: { jitsiTrack }
+        } = this.state;
+
+        return jitsiTrack && jitsiTrack.dispose();
+    }
+
+    /**
+     * Updates the current microphone data.
+     * Disposes previously created track and creates a new one.
+     *
+     * @returns {void}
+     */
+    async _updateCurrentMicData() {
+        await this._disposeCurrentTrack();
+
+        const currentMicData = await createLocalAudioTrack(
+            this.props.currentMicDeviceId,
+        );
+
+        this.setState({
```

Comment:
I suspect there's a chance of leaking audio tracks if component will unmount, before `setState` is called. Can you add artificial delay to 'createLocalAudioTrack' and see if that's the case? If it is, then this needs to be fixed, probably by doing dispose here in case the component's been unmounted already.

Reply:
This was a great feedback, thanks! Didn't think of it.